### PR TITLE
Unify runtime and setup configs in codec adapter

### PR DESCRIPTION
**Pull request by ranj063**
Rebased now.

### CODE REVIEW

**Comment by ranj063** [comment]
SOFCI TEST

**Comment by markbartonxperi** [comment]
@ranj063 Changes to DTS code look good. 
One question: do all the pipeline scripts also need to be changed as a part of this patch? Feels like `CA_SETUP_CONTROLBYTES` (plus other associated defines) should now be removed and `CA_RUNTIME_CONTROLBYTES` (plus other associated defines) should be renamed.

**Comment by ranj063** [comment]
> @ranj063 Changes to DTS code look good. One question: do all the pipeline scripts also need to be changed as a part of this patch? Feels like `CA_SETUP_CONTROLBYTES` (plus other associated defines) should now be removed and `CA_RUNTIME_CONTROLBYTES` (plus other associated defines) should be renamed.

@markbartonxperi Thanks for your feedback. Yes, the m4 files will need to be modified to make sure that the setup bytes are removed and runtime bytes renamed.

**Comment by dbaluta** [comment]
@ranj063 allow me some time to run some tests for this in the evening.

**Comment by dbaluta** [comment]
@ranj063 I might be missing something but on my machine this PR doesn't compile.

```

[ 47%] Building C object CMakeFiles/sof.dir/src/lib/notifier.c.o
[ 47%] Building C object CMakeFiles/sof.dir/src/lib/pm_runtime.c.o
In file included from /home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec/cadence.c:14:
/home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec/cadence.c: In function 'cadence_codec_init':
/home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec/cadence.c:102:42: error: 'struct module_data' has no member named 'id'
  102 |  uint32_t api_id = CODEC_GET_API_ID(codec->id);
      |                                          ^~
/home/daniel/work/sof_dir/sof/src/include/sof/audio/codec_adapter/codec/generic.h:21:32: note: in definition of macro 'CODEC_GET_API_ID'
   21 | #define CODEC_GET_API_ID(id) ((id) & 0xFF)
      |                                ^~
[ 48%] Building C object CMakeFiles/sof.dir/src/lib/clk.c.o
/home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec/cadence.c: In function 'cadence_codec_get_samples':
/home/daniel/work/sof_dir/sof/src/audio/codec_adapter/codec/cadence.c:333:42: error: 'struct module_data' has no member named 'id'
  333 |  uint32_t api_id = CODEC_GET_API_ID(codec->id);
      |                                          ^~
/home/daniel/work/sof_dir/sof/src/include/sof/audio/codec_adapter/codec/generic.h:21:32: note: in definition of macro 'CODEC_GET_API_ID'
   21 | #define CODEC_GET_API_ID(id) ((id) & 0xFF)
      |                                ^~
[ 48%] Building C object CMakeFiles/sof.dir/src/lib/dma.c.o
make[2]: *** [CMakeFiles/sof.dir/build.make:941: CMakeFiles/sof.dir/src/audio/codec_adapter/codec/cadence.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [CMakeFiles/Makefile2:1821: CMakeFiles/sof.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

**Comment by ranj063** [comment]
> @ranj063 I might be missing something but on my machine this PR doesn't compile.

@dbaluta like we discussed, this error is because I have remvoed codec_id completely. I have introduce a new field called api_id in the cadence_codec_data and set it to default to 0x01. This will avoid your error. but if you want to test the other cadecne codecs, we need a mechanism to overwrite this from userspace. Looking forward to that PR from you.

**Review by cujomalainey — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@ranj063 can you check CI, looks like a panic()

**Comment by mwasko** [comment]
> > @ranj063 can you check CI, looks like a panic()
> 
> @lgirdwood I think this is because I've removed the setup config portions and the CI tests need to be modified accordingly. @mwasko who can help check this?

@wszypelt, @kkarask can you take a look at CI tests and check how we can align them with proposed codec_adapter changes?

**Comment by dbaluta** [comment]
@ranj063 as discussed on Discord this also breaks i.MX codec adapter MP3 tests.

**Comment by kkarask** [comment]
We will check it from CI perspective together with test developer.

**Comment by kkarask** [comment]
In test 25_00_TestGenericProcessorSimplePlb48000Hz32b32b4ch I got an error from FW:
```
[     1473102.833333] (           9.500000) c0 dai          1.2            src/audio/dai.c:353  INFO dai_playback_params() fifo 0x77010
[     1473116.875000] (          14.041667) c0 pipe         1.5   ....../pipeline-params.c:308  INFO pipe prepare
[     1473217.083333] (         100.208336) c0 cadence_codec 1.0   ......../codec/generic.c:128  ERROR module_allocate_memory: requested allocation of 0 bytes.
[     1473227.833333] (          10.750000) c0 cadence_codec 1.0   ......../codec/cadence.c:291  ERROR init_memory_tables() error 0: failed to allocate memory for 2
[     1473263.833333] (          36.000000) c0 cadence_codec 1.0   ......../codec/cadence.c:448  ERROR cadence_codec_prepare() error ffffffea: failed to init memory tables
[     1473281.916667] (          18.083334) c0 cadence_codec 1.0   ......../codec/generic.c:207  ERROR module_prepare() error -22: module specific prepare failed, comp_id 0
[     1473293.000000] (          11.083333) c0 cadence_codec 1.0   ......../codec_adapter.c:133  ERROR codec_adapter_prepare() error ffffffea: codec prepare failed
[     1473304.208333] (          11.208333) c0 pipe         1.5   ....../pipeline-params.c:315  ERROR pipeline_prepare(): ret = -5, dev->comp.id = 1
[     1473315.083333] (          10.875000) c0 ipc                  src/ipc/ipc3/handler.c:292  ERROR ipc: pipe 1 comp 1 prepare failed -5
[     1473323.750000] (           8.666667) c0 pipe         1.5   ......./pipeline-graph.c:341  INFO pipe reset
[     1473369.541667] (          45.791668) c0 dai          1.2            src/audio/dai.c:671  INFO dai_reset()
```
What needs to be modified to handle this?

**Comment by lgirdwood** [comment]
@kkarask thanks, please let us know when you are good to update CI and then we can merge. Thanks !

**Comment by ranj063** [comment]
@dbaluta I have removed the ca_config parts from the topology config data for the cadence codecs now. Could you please give this PR a try again?

**Comment by ranj063** [comment]
> What needs to be modified to handle this?

@kkarask in order to handle this change the config data that is sent during the creating of the cadence codec needs to be modifed to remove the 20 bytes that contain the module ID, reserved field, sample_rate, sample_width and channels info. Have look at this commit for example: https://github.com/thesofproject/sof/pull/5190/commits/f5e062c28a3b94af87ed22d411188a5447794bc6

**Comment by ranj063** [comment]
SOFCI TEST

**Comment by dbaluta** [comment]
@ranj063  tested again, as we discussed, PR looks fine for Cadence mp3 using compress interface usecase.

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
I see this errors when a stream is closed:

```
[   381361024.221074] (        1977.656128) c0 cadence_codec 1.1   ......../codec/cadence.c:223  ERROR cadence_codec_apply_config() error: no config available
[   381361167.033569] (         142.812500) c0 cadence_codec 1.1   ......../codec/cadence.c:431  ERROR cadence_codec_prepare() error fffffffb: failed to apply config
[   381361288.752314] (         121.718742) c0 cadence_codec 1.1   ......../codec/cadence.c:564  ERROR cadence_codec_reset() error fffffffb: could not re-prepare codec after reset
[   381361411.877309] (         123.124992) c0 cadence_codec 1.1   ......../codec/generic.c:296  ERROR module_reset() error -5: module specific reset() failed for comp 1
[   381361539.116887] (         127.239578) c0 cadence_codec 1.1   ......../codec_adapter.c:553  ERROR codec_adapter_reset(): error -5, codec reset has failed
```

Let me investigate and see if there is anything wrong.

**Comment by kkarask** [comment]
CI tests are ready for this change. I've tested last change 2325fe78db67815c86d4fc8c696f924e2cdf8877 and it works.

File: src/audio/codec_adapter/codec/generic.c [hunk 1]
```diff
@@ -263,11 +263,13 @@ int module_process(struct comp_dev *dev)
 	md->state = MODULE_PROCESSING;
 
 	ret = md->ops->process(dev);
-	if (ret)
+	if (ret && ret != -ENOSPC) {
```
Comment (mwasko) [inline src/audio/codec_adapter/codec/generic.c:246]: @ranj063 can you provide more details why it should be skipped?

Comment (dbaluta) [reply src/audio/codec_adapter/codec/generic.c:246]: @mwasko ENOSPC is marker telling us that process() wasn't actually called due to not enough data in the input buffer or not not enough free space in the output buffer. This is not an error per se. It is just a state in the pipeline where we should move to next iteration in order to get more data at the input or consume more data at the output.

**Comment by ranj063** [comment]
@cujomalainey I've update this patchset to fix the cadence init and reset sequence a bit. Could you please take a look again?

**Comment by dbaluta** [comment]
@ranj063 as discussed on discord after your latest pull reset errors are gone and mp3 tests are working fine.

**Review by dbaluta — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kkarask I will sync with you and merge tomorrow morning Europe time so you can make the change to CI at the same time ?

**Comment by kkarask** [comment]
I've tested again all passed. I'm ready to change CI at the same time..

**Comment by lgirdwood** [comment]
CI run skipped some tests, need to revert so CI rerun can be full and complete all tests. CAn be applied again later if CI is good. @kkarask pls ping when we are good.

**Pull request merged**